Change completion callback param to Result<T, Error>

momo-ozawa opened the pull request:
## Description

Resolves https://github.com/wordpress-mobile/WordPress-iOS/issues/17018

This PR updates the completion callback param to `Result<T, Error>`

- Added `PurchaseError`
- Added error handling for AppStoreService.fetchProducts via `SKProductsRequestDelegate` (when `
request(_:didFailWithError:) ` gets called)
- Added error handling for AppStoreService.purchaseProduct via `SKPaymentTransactionObserver` (when the updated transaction state is `.failed`)

## Notes

- InAppPurchasesService error handling tests already exist
  - `testFetchProductSkus_WhenRequestFails_PublishesReceivedError()`
  - `testCreateOrder_WhenRequestFails_PublishesReceivedError()`

## How to test

### Running tests
1. Run MobilePayKitTests test target
2. ✅ Tests should pass

### Running the app

1. Configure the bearer token in `ProductListViewModel`
2. Run the iOS app
3. Purchase the product by tapping on it
4. Tap Confirm
5. Tap OK
6. On Xcode debugger menu, click on the Manage StoreKit Transactions menu item 
7. ✅ There should be a record of a product being purchased (See example below)

<img width="747" alt="Screen Shot 2021-07-19 at 16 32 32" src="https://user-images.githubusercontent.com/6711616/126186582-3f76559e-7fe5-45ae-9068-2754883f6351.png">

Commit:
Refactor: change fetch closure param to Result

Commit:
Add: SKRequestDelegate didFailWithError method

Commit:
Refactor: propagate fetch product skus error

Commit:
Refactor: change purchase closure param to Result

Commit:
Refactor: propagate create order error

Commit:
Refactor: propagate transaction state failed error

Commit:
Refactor: update completion param to Result type

Commit:
Fix: typo

Commit:
Refactor: use InAppPurchasesService protocol as a type

Commit:
Add: purchase callback helper method

Commit:
Add: PurchaseError enum

Commit:
Add: fetch callback helper method

leandroalonso reviewed (dismissed):
`testPaymentQueueUpdatedTransactions_WhenTransactionStateIsFailed_CallsFinishTransaction` it's failing for me (running on `iPhone 11 14.5`)

There's any way to simulate a purchase error?

nikola-milicevic commented on Source/Services/AppStoreService.swift:
```diff
@@ -53,15 +53,15 @@ public class AppStoreService: NSObject, AppStoreServiceProtocol {
 
     public func fetchProducts(completion: @escaping FetchCompletionCallback) {
-        iapService.fetchProductSKUs()
+        iapService.fetchProductSkus()
```
I would keep the `fetchProductSKUs` version, I think it goes more inline with the rest of the SDK, like `url.absoluteURL` for example.

momo-ozawa replied:
Fixed! 5fa3bbb

nikola-milicevic commented on Example/iOS/MobilePayKit-iOS/ProductListViewModel.swift:
```diff
@@ -33,17 +33,24 @@ class ProductListViewModel: ObservableObject {
+                self.products = products.map { Product(product: $0) }
+            case .failure(let error):
+                print("Error fetching products: \(error.localizedDescription)")
+            }
```
I would:

1. use the trailing closure
2. omit `result` and use $0 syntax
3.  maybe add `weak self`, not sure if it's needed here though

Result:

```
MobilePayKit.shared.fetchProducts { [weak self] in
    switch $0 {
    case .success(let products):
        self?.products = products.map { Product(product: $0) }
    case .failure(let error):
        print("Error fetching products: \(error.localizedDescription)")
    }
}
```

momo-ozawa replied:
> use the trailing closure

Fixed! ecbf98f

> maybe add weak self, not sure if it's needed here though

Fixed! b8e72a1

> omit result and use $0 syntax

I decided to keep `result`, because I think it makes it more readable in this case.

nikola-milicevic commented on Source/Services/AppStoreService.swift:
```diff
@@ -2,8 +2,8 @@ import Combine
-public typealias FetchCompletionCallback = ([SKProduct]) -> Void
-public typealias PurchaseCompletionCallback = (SKPaymentTransaction?) -> Void
+public typealias FetchCompletionCallback = (Result<[SKProduct], Error>) -> Void
+public typealias PurchaseCompletionCallback = (Result<SKPaymentTransaction, Error>) -> Void
```
I would maybe go even further and define:

```
public typealias ProductResult = Result<[SKProduct], Error>
public typealias TransactionResult = Result<SKPaymentTransaction, Error>

```
and use it here and in the rest of the code, to avoid repeating.

momo-ozawa replied:
Fixed! b62cbcd

nikola-milicevic reviewed (commented):
Great work, I just left some suggestions about the style.

Commit:
Refactor: use typealias for ProductsResult and TransactionResult

Commit:
Refactor: capitalize SKU

Commit:
Fix: use trailing closure

Use a trailing closure if a function has a single closure argument and it is the final argument

Commit:
Fix: line wrapping

Commit:
Fix: weak self

Commit:
Fix: payment queue txn failed test

momo-ozawa commented:
> `testPaymentQueueUpdatedTransactions_WhenTransactionStateIsFailed_CallsFinishTransaction` it's failing for me (running on `iPhone 11 14.5`)
> 
> There's any way to simulate a purchase error?

@leandroalonso Fixed the test! 418aca8

momo-ozawa commented:
@leandroalonso @nikola-milicevic 
Thanks for reviewing! Ready for a re-review 😄 🙇‍♀️

Commit:
Fix: skus -> SKUs

nikola-milicevic reviewed (approved):
Tests green, looks good! 🚀
